### PR TITLE
Akash update leverj futures20200831

### PR DESCRIPTION
**Pull request by mitakash**

### CODE REVIEW

File: market_maker_keeper/leverjfutures_market_maker_keeper.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        if (quote == product['quoteSymbol']):
+            if ((target_price is None) or (target_price.buy_price is None) or (target_price.sell_price is None)):
+                quote_allocation = Wad(0)
+            else:
+                quote_allocation = Wad(int(float(total_available)/(total_number_of_instruments*2)))
+            self.logger.info(f'quote_allocation: {quote_allocation}')
+        allocation = {base: net_base_allocation, quote: quote_allocation}
+        return allocation
+
+    def _exponent_for_lowest_denomination(self, coin: str) -> int:
```
Comment (ith-harvey) [inline market_maker_keeper/leverjfutures_market_maker_keeper.py:218]: I'm a little confused on this one? This method will always return the int 18

File: market_maker_keeper/leverjfutures_market_maker_keeper.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            if ((target_price is None) or (target_price.buy_price is None) or (target_price.sell_price is None)):
+                net_base_allocation = Wad(0)
+            else:
+                average_price = (int(target_price.buy_price) + int(target_price.sell_price))/2
+                self.logger.info(f'target_price, average_price: {average_price}')
+                conversion_divisor = int(average_price*total_number_of_instruments*2*buffer_adjustment_factor)
+                exponent = self._exponent_for_lowest_denomination(base)
+                base_allocation = (float(total_available)/(conversion_divisor*10**(18 - exponent)))
+                base_allocation_wad = Wad(int(base_allocation))
+                open_position_for_base = self.current_open_position(base)
```
Comment (ith-harvey) [inline market_maker_keeper/leverjfutures_market_maker_keeper.py:205]: this line could be refactored to:

`open_position_for_base = self.leverj_api.get_position_in_wad(base)`

File: market_maker_keeper/leverjfutures_market_maker_keeper.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    def main(self):
+        with Lifecycle() as lifecycle:
+            lifecycle.initial_delay(1)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(1, self.synchronize_orders)
+            lifecycle.on_shutdown(self.shutdown)
+
+    def startup(self):
+        #quote_increment = 1/(self.leverj_api.get_product(self.arguments.pair)["ticksperpoint"])
+        quote_increment = self.leverj_api.get_product(self.pair())["tickSize"]
```
Comment (ith-harvey) [inline market_maker_keeper/leverjfutures_market_maker_keeper.py:163]: I would actually abstract this method and put it into pyexchange...

This is what we do for other MM keepers - https://github.com/makerdao/pyexchange/blob/75f4bccf02c0b0d1e5cb4bd8b62d652a75ab93c5/pyexchange/binance_us.py#L83

File: market_maker_keeper/leverjfutures_market_maker_keeper.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        return self.arguments.pair.upper()[:3]
+
+    def token_buy(self) -> str:
+        return self.arguments.pair.upper()[3:]
+
+
+    def allocated_balance(self, our_balances: dict, token: str) -> Wad:
+        quote_asset_address = self.leverj_api.get_product(self.pair())["quote"]["address"]
+        total_available = our_balances[quote_asset_address]['available']
+        self.logger.info(f'total_available: {total_available}')
```
Comment (ith-harvey) [inline market_maker_keeper/leverjfutures_market_maker_keeper.py:183]: Make this a debug log instead?

File: market_maker_keeper/leverjfutures_market_maker_keeper.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        return self._allocate_to_pair(total_available).get(token)
+
+    def _allocate_to_pair(self, total_available):
```
Comment (ith-harvey) [inline market_maker_keeper/leverjfutures_market_maker_keeper.py:186]: Could you add more comments throughout detailing what different variables are? Would be good to provide some more context here.

not sure what `conversion_divisor` and `buffer_adjustment_factor` are

File: market_maker_keeper/leverjfutures_market_maker_keeper.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        buffer_adjustment_factor = 1.1
+        base = self.arguments.pair.upper()[:3]
+        quote = self.arguments.pair.upper()[3:]
+        target_price = self.price_feed.get_price()
+        product = self.leverj_api.get_product(self.pair())
+        if (base == product['baseSymbol']):
+            if ((target_price is None) or (target_price.buy_price is None) or (target_price.sell_price is None)):
+                net_base_allocation = Wad(0)
+            else:
+                average_price = (int(target_price.buy_price) + int(target_price.sell_price))/2
```
Comment (ith-harvey) [inline market_maker_keeper/leverjfutures_market_maker_keeper.py:199]: So here and below you seem to do alot of `int` and `float` wrapping on multiple lines. Just wanted to ask here if you tested reducing or combining some of these to clean things up?

File: market_maker_keeper/leverjfutures_market_maker_keeper.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+                base_allocation_wad = Wad(int(base_allocation))
+                open_position_for_base = self.current_open_position(base)
+                calculated_net_allocation = base_allocation_wad.value + open_position_for_base.value
+                net_base_allocation = Wad(calculated_net_allocation) if calculated_net_allocation > 0 else Wad(0)
+            self.logger.info(f'net_base_allocation: {net_base_allocation}')
+        if (quote == product['quoteSymbol']):
+            if ((target_price is None) or (target_price.buy_price is None) or (target_price.sell_price is None)):
+                quote_allocation = Wad(0)
+            else:
+                quote_allocation = Wad(int(float(total_available)/(total_number_of_instruments*2)))
```
Comment (ith-harvey) [inline market_maker_keeper/leverjfutures_market_maker_keeper.py:213]: so this is scary for me... don't know if you intended this...

but taking a `float `and wrapping it in an `int` can alter the inner value:
```
my_int = 2
my_float = 2.2

output = int(float(my_int))
float_output = int(float(my_float))

print(output)
print(float_output)
```
In the above example both print `2`. Was this intended?

**Review by mitakash — COMMENTED** [review]
most of these old comments from Ian don't apply anymore, Ian needs to look at the new code, because there is new logic and review.

**Review by ith-harvey — COMMENTED** [review]
updated request changes

**Review by ith-harvey — APPROVED** [review]

**Pull request merged**